### PR TITLE
Tabular: Added internal support for _FULL ensembles

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

Added internal support for _FULL ensemble creation that only uses other _FULL ensembles during prediction. This results in 10-200x inference time speedup for most ensembles with minimal accuracy loss.

Once this PR is merged, the follow-up PR will introduce this to Tabular Predictor as a user-facing option. For that PR, I will also benchmark the functionality to ensure stability and demonstrate the difference in accuracy and inference time.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
Example code to showcase functionality:


```
import pandas as pd
import sklearn.datasets

from autogluon.utils.tabular.ml.utils import generate_train_test_split
from autogluon import TabularPrediction as task

LABEL = 'target'


if __name__ == '__main__':
    cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
    X = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
    X[LABEL] = cancer_data[LABEL]

    X_train, X_test, y_train, y_test = generate_train_test_split(X=X.drop(LABEL, axis=1), y=X[LABEL], problem_type='binary', test_size=0.2)

    X = X_train
    X[LABEL] = y_train
    X_test[LABEL] = y_test

    X = X.reset_index(drop=True)
    X_test = X_test.reset_index(drop=True)

    predictor_stack = task.fit(
        time_limits=200,
        train_data=X, label=LABEL,
        hyperparameters={
            'GBM': {'num_boost_round': 10},
            'CAT': {'iterations': 10},
            'KNN': {},
        },
        num_bagging_folds=10, stack_ensemble_levels=2,
        num_bagging_sets=2,
    )

    predictor_stack._learner.refit_ensemble_full()

    predictor_stack.leaderboard()
    predictor_stack.leaderboard(X_test)

```

Example Output:

```
AutoGluon training complete, total runtime = 49.48s ...
Fitting model: KNeighborsClassifierUnif_FULL_STACKER_l0 ...
	0.02s	 = Training runtime
Fitting model: KNeighborsClassifierDist_FULL_STACKER_l0 ...
	0.02s	 = Training runtime
Fitting model: LightGBMClassifier_FULL_STACKER_l0 ...
	0.18s	 = Training runtime
Fitting model: CatboostClassifier_FULL_STACKER_l0 ...
	0.09s	 = Training runtime
Fitting model: KNeighborsClassifierUnif_FULL_STACKER_l1 ...
	0.03s	 = Training runtime
Fitting model: KNeighborsClassifierDist_FULL_STACKER_l1 ...
	0.02s	 = Training runtime
Fitting model: LightGBMClassifier_FULL_STACKER_l1 ...
	0.16s	 = Training runtime
Fitting model: CatboostClassifier_FULL_STACKER_l1 ...
	0.12s	 = Training runtime
Fitting model: weighted_ensemble_FULL_k0_l1 ...
	0.9736	 = Validation accuracy score
	0.41s	 = Training runtime
	0.0s	 = Validation runtime
Fitting model: KNeighborsClassifierUnif_FULL_STACKER_l2 ...
	0.02s	 = Training runtime
Fitting model: KNeighborsClassifierDist_FULL_STACKER_l2 ...
	0.03s	 = Training runtime
Fitting model: LightGBMClassifier_FULL_STACKER_l2 ...
	0.16s	 = Training runtime
Fitting model: CatboostClassifier_FULL_STACKER_l2 ...
	0.11s	 = Training runtime
Fitting model: weighted_ensemble_FULL_k0_l2 ...
	0.9758	 = Validation accuracy score
	0.3s	 = Training runtime
	0.0s	 = Validation runtime
Fitting model: weighted_ensemble_FULL_k0_l3 ...
	0.9714	 = Validation accuracy score
	0.32s	 = Training runtime
	0.0s	 = Validation runtime
                                       model  score_val  pred_time_val   fit_time  pred_time_val_marginal  fit_time_marginal  stack_level
0                    weighted_ensemble_k0_l2   0.975824      10.294425  19.603538                0.001510           0.409049            2
1              CatboostClassifier_STACKER_l0   0.973626       0.191732   3.356845                0.191732           3.356845            0
2                    weighted_ensemble_k0_l1   0.973626       0.193289   3.680599                0.001557           0.323754            1
3              LightGBMClassifier_STACKER_l1   0.971429       7.497031  15.285516                0.296525           4.005262            1
4                    weighted_ensemble_k0_l3   0.971429      13.298196  27.130733                0.001614           0.329166            3
5              CatboostClassifier_STACKER_l1   0.969231       7.441637  14.654596                0.241132           3.374342            1
6              CatboostClassifier_STACKER_l2   0.969231      12.976022  22.937808                0.224354           3.258199            2
7              LightGBMClassifier_STACKER_l2   0.969231      13.072228  23.543368                0.320560           3.863760            2
8              LightGBMClassifier_STACKER_l0   0.951648       0.385332   5.391079                0.385332           5.391079            0
9        KNeighborsClassifierUnif_STACKER_l2   0.945055      15.239536  20.168296                2.487868           0.488688            2
10       KNeighborsClassifierDist_STACKER_l2   0.945055      15.284509  20.194434                2.532841           0.514826            2
11       KNeighborsClassifierUnif_STACKER_l0   0.940659       3.582203   1.539624                3.582203           1.539624            0
12       KNeighborsClassifierDist_STACKER_l1   0.940659       9.659258  11.765373                2.458753           0.485119            1
13       KNeighborsClassifierUnif_STACKER_l1   0.938462       9.755258  11.814885                2.554752           0.534631            1
14       KNeighborsClassifierDist_STACKER_l0   0.936264       3.041238   0.992705                3.041238           0.992705            0
15              weighted_ensemble_FULL_k0_l3        NaN            NaN   1.226404                     NaN           0.317221            3
16              weighted_ensemble_FULL_k0_l2        NaN            NaN   0.917830                     NaN           0.299806            2
17              weighted_ensemble_FULL_k0_l1        NaN            NaN   0.496346                     NaN           0.405361            1
18        LightGBMClassifier_FULL_STACKER_l2        NaN            NaN   0.796874                     NaN           0.160787            2
19        LightGBMClassifier_FULL_STACKER_l1        NaN            NaN   0.461048                     NaN           0.157392            1
20        LightGBMClassifier_FULL_STACKER_l0        NaN            NaN   0.180238                     NaN           0.180238            0
21  KNeighborsClassifierUnif_FULL_STACKER_l2        NaN            NaN   0.656484                     NaN           0.020397            2
22  KNeighborsClassifierUnif_FULL_STACKER_l1        NaN            NaN   0.335883                     NaN           0.032227            1
23  KNeighborsClassifierUnif_FULL_STACKER_l0        NaN            NaN   0.016244                     NaN           0.016244            0
24  KNeighborsClassifierDist_FULL_STACKER_l2        NaN            NaN   0.664533                     NaN           0.028446            2
25  KNeighborsClassifierDist_FULL_STACKER_l1        NaN            NaN   0.321719                     NaN           0.018063            1
26  KNeighborsClassifierDist_FULL_STACKER_l0        NaN            NaN   0.016189                     NaN           0.016189            0
27        CatboostClassifier_FULL_STACKER_l2        NaN            NaN   0.748396                     NaN           0.112309            2
28        CatboostClassifier_FULL_STACKER_l1        NaN            NaN   0.428405                     NaN           0.124749            1
29        CatboostClassifier_FULL_STACKER_l0        NaN            NaN   0.090985                     NaN           0.090985            0
                                       model  score_test  score_val  pred_time_test  pred_time_val   fit_time  pred_time_test_marginal  pred_time_val_marginal  fit_time_marginal  stack_level
0         CatboostClassifier_FULL_STACKER_l0    0.973684        NaN        0.010251            NaN   0.090985                 0.010251                     NaN           0.090985            0
1               weighted_ensemble_FULL_k0_l1    0.973684        NaN        0.015546            NaN   0.496346                 0.005295                     NaN           0.405361            1
2         LightGBMClassifier_FULL_STACKER_l1    0.973684        NaN        0.295506            NaN   0.461048                 0.013881                     NaN           0.157392            1
3         CatboostClassifier_FULL_STACKER_l2    0.973684        NaN        0.604655            NaN   0.748396                 0.017777                     NaN           0.112309            2
4         LightGBMClassifier_FULL_STACKER_l2    0.973684        NaN        0.617470            NaN   0.796874                 0.030592                     NaN           0.160787            2
5               weighted_ensemble_FULL_k0_l3    0.973684        NaN        0.640332            NaN   1.226404                 0.005085                     NaN           0.317221            3
6         CatboostClassifier_FULL_STACKER_l1    0.964912        NaN        0.302256            NaN   0.428405                 0.020631                     NaN           0.124749            1
7               weighted_ensemble_FULL_k0_l2    0.964912        NaN        0.463235            NaN   0.917830                 0.007720                     NaN           0.299806            2
8              CatboostClassifier_STACKER_l1    0.964912   0.969231        4.818498       7.441637  14.654596                 0.063165                0.241132           3.374342            1
9                    weighted_ensemble_k0_l2    0.964912   0.975824        7.374699      10.294425  19.603538                 0.005218                0.001510           0.409049            2
10             CatboostClassifier_STACKER_l2    0.964912   0.969231        9.660040      12.976022  22.937808                 0.082540                0.224354           3.258199            2
11             CatboostClassifier_STACKER_l0    0.956140   0.973626        0.077349       0.191732   3.356845                 0.077349                0.191732           3.356845            0
12                   weighted_ensemble_k0_l1    0.956140   0.973626        0.079961       0.193289   3.680599                 0.002612                0.001557           0.323754            1
13             LightGBMClassifier_STACKER_l1    0.956140   0.971429        5.049974       7.497031  15.285516                 0.294641                0.296525           4.005262            1
14             LightGBMClassifier_STACKER_l2    0.956140   0.969231        9.853971      13.072228  23.543368                 0.276472                0.320560           3.863760            2
15                   weighted_ensemble_k0_l3    0.956140   0.971429        9.939996      13.298196  27.130733                 0.003484                0.001614           0.329166            3
16        LightGBMClassifier_FULL_STACKER_l0    0.947368        NaN        0.021221            NaN   0.180238                 0.021221                     NaN           0.180238            0
17             LightGBMClassifier_STACKER_l0    0.947368   0.951648        0.279679       0.385332   5.391079                 0.279679                0.385332           5.391079            0
18  KNeighborsClassifierDist_FULL_STACKER_l0    0.912281        NaN        0.122854            NaN   0.016189                 0.122854                     NaN           0.016189            0
19  KNeighborsClassifierUnif_FULL_STACKER_l0    0.912281        NaN        0.127299            NaN   0.016244                 0.127299                     NaN           0.016244            0
20  KNeighborsClassifierDist_FULL_STACKER_l1    0.912281        NaN        0.412988            NaN   0.321719                 0.131363                     NaN           0.018063            1
21  KNeighborsClassifierUnif_FULL_STACKER_l1    0.912281        NaN        0.421003            NaN   0.335883                 0.139378                     NaN           0.032227            1
22  KNeighborsClassifierDist_FULL_STACKER_l2    0.912281        NaN        0.724014            NaN   0.664533                 0.137136                     NaN           0.028446            2
23  KNeighborsClassifierUnif_FULL_STACKER_l2    0.912281        NaN        0.745266            NaN   0.656484                 0.158388                     NaN           0.020397            2
24       KNeighborsClassifierUnif_STACKER_l0    0.912281   0.940659        2.195365       3.582203   1.539624                 2.195365                3.582203           1.539624            0
25       KNeighborsClassifierDist_STACKER_l0    0.912281   0.936264        2.202940       3.041238   0.992705                 2.202940                3.041238           0.992705            0
26       KNeighborsClassifierDist_STACKER_l1    0.912281   0.940659        6.963352       9.659258  11.765373                 2.208019                2.458753           0.485119            1
27       KNeighborsClassifierUnif_STACKER_l1    0.912281   0.938462        7.011675       9.755258  11.814885                 2.256342                2.554752           0.534631            1
28       KNeighborsClassifierUnif_STACKER_l2    0.912281   0.945055       11.749548      15.239536  20.168296                 2.172048                2.487868           0.488688            2
29       KNeighborsClassifierDist_STACKER_l2    0.912281   0.945055       11.755782      15.284509  20.194434                 2.178282                2.532841           0.514826            2

```

**Comment by szha** [comment]
Job PR-408-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-408/1/index.html

**Comment by szha** [comment]
Job PR-408-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-408/2/index.html

File: autogluon/utils/tabular/ml/constants.py [hunk 1]
```diff
@@ -10,3 +10,4 @@
 
 REFIT_FULL_NAME = 'refit_single_full'  # stack-name used for refit_single_full (aka "compressed") models
 REFIT_FULL_SUFFIX = "_FULL"  # suffix appended to model name for refit_single_full (aka "compressed") models
+REFIT_FULL_PREFIX = "FULL_"
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/constants.py:13]: why is this prefix sometimes needed? (ie. why is suffix insufficient?)

File: autogluon/utils/tabular/ml/learner/abstract_learner.py [hunk 2]
```diff
@@ -172,6 +172,11 @@ def refit_single_full(self, models=None):
         trainer = self.load_trainer()
         return trainer.refit_single_full(X=X, y=y, models=models)
 
+    # Fits _FULL models and links them in the stack so _FULL models only use other _FULL models as input during stacking
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/learner/abstract_learner.py:175]: Is the reason for separate `refit_single_full` vs `refit_ensemble_full` methods is because the former just adds one extra new model whereas the latter may modify many models?
Otherwise, I could imagine `refit_single_full` just being a special case of the `refit_ensemble_full` method (ie. only exposing one method to user to apply to both ensembles and bagged base models, and even non-bagged base models).

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py [hunk 3]
```diff
@@ -104,6 +103,7 @@ def __init__(self, path: str, problem_type: str, scheduler_options=None, objecti
         self.model_pred_times = {}
         self.models = {}
         self.model_graph = nx.DiGraph()
+        self.model_compressed_dict = {}  # Dict of Uncompressed Model -> Compressed Model
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/trainer/abstract_trainer.py:106]: we'll need to standardize terminology between "FULL" and "COMPRESSED".  I prefer "FULL" as it more clearly warns against interpreting accuracies on any of the provided data.

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py [hunk 4]
```diff
@@ -806,37 +798,122 @@ def get_inputs_to_stacker(self, X, level_start, level_end, model_levels=None, y_
...
+        for model in model_compressed_dict.keys():
+            if model_compressed_dict[model] not in models_trained_full:
+                keys_to_del.append(model)
+        for key in keys_to_del:
+            del model_compressed_dict[key]
+        self.model_compressed_dict.update(model_compressed_dict)
+        self.save()  # TODO: This could be more efficient by passing in arg to not save if called by refit_ensemble_full since it saves anyways later.
         return models_trained_full
 
+    # TODO: Rename _FULL -> _SINGLE / _SOLO?
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/trainer/abstract_trainer.py:878]: I think FULL actually sounds ok, since the biggest concern is that users realize the model has been trained on all of their data so they cannot view it's accuracies as unbiased.

File: autogluon/utils/tabular/ml/learner/abstract_learner.py (hunk 2 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/learner/abstract_learner.py:175]: refit_single_full trains a model _FULL without altering its input sources, so a stacker would still use the non-full version models as input. refit_ensemble_full calls refit_single_full, then alters the _FULL models to take other _FULL models as input instead of the normal models. I will try to clarify in documentation.

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py (hunk 3 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/trainer/abstract_trainer.py:106]: Agreed, changed all references to compressed to full

File: autogluon/utils/tabular/ml/constants.py (hunk 1 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/constants.py:13]: Removed, this was a minor hack that wasn't necessary.

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py (hunk 4 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/trainer/abstract_trainer.py:878]: Agreed, removed TODO

**Comment by szha** [comment]
Job PR-408-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-408/3/index.html
